ratis-examples/src/test/java/org/apache/ratis/examples/ParameterizedBaseTest.java

```diff
@@ -123,6 +126,11 @@ public boolean hasNext() {
     final RaftProperties prop = new RaftProperties();
     prop.setClass(MiniRaftCluster.STATEMACHINE_CLASS_KEY,
         stateMachineClass, StateMachine.class);
+
+    // make sure leadership check won't affect the test
+    RaftServerConfigKeys.Rpc.setTimeoutMin(prop, TimeDuration.valueOf(2, TimeUnit.SECONDS));
+    RaftServerConfigKeys.Rpc.setTimeoutMax(prop, TimeDuration.valueOf(4, TimeUnit.SECONDS));
```

Comment:
We should not ideally require setting timeout in the tests. Are we experiencing flakiness in the tests due to the change?

Reply:
Yes, some test assume that leader will not step down even it can not hear from followers. 

Such as `runTestUpdateViaHeartbeat` in `TestRaftWithGrpc`, after leader is elected, it just block followers's append requests for 5 seconds, then unblock them

```
    waitForLeader(cluster);
    try (final RaftClient client = cluster.createClient()) {
      // block append requests
      cluster.getServerAliveStream()
          .filter(impl -> !impl.isLeader())
          .map(SimpleStateMachine4Testing::get)
          .forEach(SimpleStateMachine4Testing::blockWriteStateMachineData);

      CompletableFuture<RaftClientReply>
          replyFuture = client.sendAsync(new RaftTestUtil.SimpleMessage("abc"));
      TimeDuration.valueOf(5 , TimeUnit.SECONDS).sleep();
      // replyFuture should not be completed until append request is unblocked.
      Assert.assertTrue(!replyFuture.isDone());
      // unblock append request.
      cluster.getServerAliveStream()
          .filter(impl -> !impl.isLeader())
          .map(SimpleStateMachine4Testing::get)
          .forEach(SimpleStateMachine4Testing::unblockWriteStateMachineData);
```

Such case will failed due to `checkLeadership`

Reply:
blockWriteStateMachineData - it only blocks StateMachine$DataApi#write call. Heartbeats should still work.

Reply:
appendEntries in Follower is blocked by StateMachine$DataApi#write, thus won't reply append entries response, which will cause leader to step down.

Reply:
Hi @lokeshj1703 
raft use bi-direction streaming grpc, it will maintain sequence between requests and sequence between responses. 

Leader do send heartbeats after sending the append entries request which will be blocked by StateMachine$DataApi#write, but it won't receive replies of these heartbeats until receiving reply of that append entries response.

This sequence assurance is important for maintain log match property of raft.